src/named_pipe.rs

```diff
@@ -0,0 +1,208 @@
+
+use crate::error::{WorkerError, WorkerErrorKind};
+
+#[derive(Debug)]
+pub struct NamedPipeCreator {
+    dir: TempDir,
+    counter: AtomicU32,
+}
+
+impl NamedPipeCreator {
+    pub fn new() -> Result<Self, WorkerError> {
+        Ok(Self {
+            dir: TempDir::new()?,
+            counter: AtomicU32::new(0),
+        })
+    }
+
+    pub fn new_pipe(&self) -> Result<NamedPipe, WorkerError> {
+        let pipe_num = self.counter.fetch_add(1, Ordering::SeqCst);
+
+        let component_input_fifo_filename = format!("IN_{}", pipe_num);
+        let component_input_fifo_path = self.dir.path().join(component_input_fifo_filename);
+
+        let component_output_fifo_filename = format!("OUT_{}", pipe_num);
+        let component_output_fifo_path = self.dir.path().join(component_output_fifo_filename);
+
+        mkfifo(
+            &component_input_fifo_path,
+            Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+        )?;
+        mkfifo(
+            &component_output_fifo_path,
+            Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+        )?;
+
+        debug!(
+            "Creating new pipe I = {:?}, O = {:?}",
+            component_input_fifo_path, component_output_fifo_path
+        );
+
+        Ok(NamedPipe {
+            component_input_fifo_path,
+            component_output_fifo_path,
+
+            component_input_fifo_file: None,
+            component_output_fifo_file: None,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct NamedPipe {
+    component_input_fifo_path: PathBuf,
+    component_output_fifo_path: PathBuf,
+
+    component_input_fifo_file: Option<File>,
+    component_output_fifo_file: Option<File>,
+}
+
+const PIPE_CREATION_TIMEOUT_MS: u64 = 100;
+const PIPE_IO_TIMEOUT_MS: u64 = 1000;
+const PIPE_POLL_INTERVAL_MS: u64 = 2;
+
+const BUF_SIZE: usize = 256;
+
+impl NamedPipe {
+    fn get_fds(&mut self) -> Result<(RawFd, RawFd), WorkerError> {
+        let deadline = Instant::now() + Duration::from_millis(PIPE_CREATION_TIMEOUT_MS);
+
+        while self.component_output_fifo_file.is_none() && Instant::now() < deadline {
+            let c_out_res = OpenOptions::new()
+                .read(true)
+                .custom_flags(OFlag::O_NONBLOCK.bits())
+                .open(&self.component_output_fifo_path);
+
+            debug!("Opening component output {:?}", c_out_res);
+
+            self.component_output_fifo_file = c_out_res.ok();
+
+            sleep(Duration::from_millis(PIPE_POLL_INTERVAL_MS))
+        }
+
+        while self.component_input_fifo_file.is_none() && Instant::now() < deadline {
+            let c_in_res = OpenOptions::new()
+                .write(true)
+                .custom_flags(OFlag::O_NONBLOCK.bits())
+                .open(&self.component_input_fifo_path);
+
+            debug!("Opening component input {:?}", c_in_res);
+
+            self.component_input_fifo_file = c_in_res.ok();
+
+            sleep(Duration::from_millis(PIPE_POLL_INTERVAL_MS))
```

Comment:
Why do we need to sleep?

Reply:
We just don't want to waste cycles -- want to give the subprocess a chance to run

The code is correct without the sleeping